crates/wysiwyg/src/dom/action_list.rs

```diff
@@ -0,0 +1,316 @@
+                DomAction::Remove(a) => to_remove.push(a),
+                DomAction::Move(a) => to_move.push(a),
+            }
+        }
+        (to_add, to_remove, to_move)
+    }
+
+    pub fn push(&mut self, action: DomAction<S>) {
+        self.actions.push(action);
+    }
+
+    #[allow(dead_code)]
+    pub fn replace_actions(&mut self, new_actions: Vec<DomAction<S>>) {
+        self.actions = new_actions;
+    }
+
+    pub fn find_moved_parent_or_self(
+        &self,
+        handle: &DomHandle,
+    ) -> Option<(DomHandle, DomHandle)> {
+        for action in &self.actions {
+            match action {
+                DomAction::Move(a) => {
+                    if a.from_handle.is_parent_of(handle)
+                        || a.from_handle == *handle
+                    {
+                        return Some((
+                            a.from_handle.clone(),
+                            a.to_handle.clone(),
+                        ));
+                    }
+                }
+                _ => (),
+            }
+        }
+        None
+    }
+}
+
+impl<S: UnicodeString> Extend<DomAction<S>> for DomActionList<S> {
+    fn extend<T: IntoIterator<Item = DomAction<S>>>(&mut self, iter: T) {
+        self.actions.extend(iter);
+    }
+}
+
+impl<S: UnicodeString> IntoIterator for DomActionList<S> {
+    type Item = DomAction<S>;
+    type IntoIter = std::vec::IntoIter<DomAction<S>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.actions.into_iter()
+    }
+}
+
+impl<S: UnicodeString> Default for DomActionList<S> {
+    fn default() -> Self {
+        Self {
+            actions: Vec::new(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use widestring::Utf16String;
+
+    type DAL = DomActionList<Utf16String>;
+
+    #[test]
+    fn default_creates_empty_action_list() {
+        let list = DAL::default();
+        assert!(list.actions.is_empty());
+    }
+
+    #[test]
+    fn new_creates_action_list_with_passed_actions() {
+        let actions = default_actions();
+        let list = DAL::new(actions.clone());
+        assert_eq!(actions, list.actions);
+    }
+
+    #[test]
+    fn new_can_create_empty_list_too() {
+        let list = DAL::new(Vec::new());
+        assert!(list.actions.is_empty());
+    }
+
+    #[test]
+    fn actions_fn_returns_all_actions() {
+        let actions = default_actions();
+        let list = DAL::new(actions.clone());
+        assert_eq!(list.actions().get(0).unwrap(), actions.get(0).unwrap());
+        assert_eq!(list.actions().get(1).unwrap(), actions.get(1).unwrap());
+        assert_eq!(list.actions().get(2).unwrap(), actions.get(2).unwrap());
+    }
+
+    #[test]
+    fn additions_returns_only_add_actions() {
+        let actions = default_actions();
+        let list = DAL::new(actions.clone());
+        assert_eq!(1, list.additions().len());
+        assert_eq!(
+            DomAction::Add(list.additions().get(0).unwrap().clone().clone()),
```

Comment:
Double clone?

Reply:
Also in a couple of other places.

Reply:
I think it's a double reference. Should I use `deref()` instead? I'm still quite inexperienced in this part of memory management in Rust 😅 .

Reply:
Should actually sort itself out if you replace `get(0).unwrap()` with `[0]`.